crates/subspace-archiving/src/archiver.rs

```diff
@@ -599,75 +726,109 @@ impl Archiver {
 
             // Push back shortened segment item
             items.push(segment_item);
+            segment_size -= spill_over;
         } else {
             // Above code added bytes length even though it was assumed that all continuation bytes
             // fit into the segment, now we need to tweak that
             last_archived_block.set_complete();
         }
 
+        // Segment is full, so we need to produce incremental mappings.
+        let new_segment_blocks = match segment.items() {
+            // If there is a BlockContinuation before this block, we need to produce its mappings
+            // as well, because they weren't produced by the previous produce_segment() or
+            // with_initial_state() call.
+            [.., SegmentItem::BlockContinuation { .. }, SegmentItem::Block { .. } | SegmentItem::BlockStart { .. }] => {
+                &segment.items()[segment.len() - 2..]
+            }
+            // Otherwise, just produce mappings for the final block, which can be continued or
+            // split.
+            [.., SegmentItem::BlockContinuation { .. }
+            | SegmentItem::Block { .. }
+            | SegmentItem::BlockStart { .. }] => &segment.items()[segment.len() - 1..],
+            // No new blocks, unreachable as long as the caller added a new block
+            // (or we're still processing a large block continuation)
+            _ => {
+                unreachable!(
+                    "Caller always adds a new block, or previous call adds block continuation; qed. {}",
+                    segment,
+                )
+            }
```

Comment:
Similar question here: are we absolutely sure there are no other valid patterns?